include/rmm/mr/device/callback_memory_resource.hpp

```diff
@@ -0,0 +1,121 @@
+ */
+#pragma once
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cstddef>
+#include <functional>
+#include <utility>
+
+namespace rmm::mr {
+
+/**
+ * @brief Callback function type used by callback memory resource for allocation.
+ *
+ * The signature of the callback function is:
+ *   `void* allocate_callback_t(std::size_t bytes, cuda_stream_view stream, void* arg);
+ *
+ * * Returns a pointer to an allocation of at least `bytes` usable immediately on
+ *   `stream`. The stream-ordered behavior requirements are identical to
+ *   `device_memory_resource::allocate`.
+ *
+ * * Additional context to the callback may be passed via `arg` that is provided at
+ *   construction of the `callback_memory_resource`.
+ */
+using allocate_callback_t = std::function<void*(std::size_t, cuda_stream_view, void*)>;
+
+/**
+ * @brief Callback function type used by callback_memory_resource for deallocation.
+ *
+ * The signature of the callback function is:
+ *   `void deallocate_callback_t(void* ptr, std::size_t bytes, cuda_stream_view stream, void* arg);
+ *
+ * * Deallocates memory pointed to by `ptr`. `bytes` specifies the size of the allocation
+ *   in bytes, and must equal the value of `bytes` that was passed to the allocate callback
+ *   function. The stream-ordered behavior requirements are identical to
+ *   `device_memory_resource::deallocate`.
+ *
+ * * Additional context to the callback may be passed via `arg` that is provided at
+ *   construction of the `callback_memory_resource`.
+ */
+using deallocate_callback_t = std::function<void(void*, std::size_t, cuda_stream_view, void*)>;
+
+/**
+ * @brief A device memory resource that uses the provided callbacks for memory allocation
+ * and deallocation.
+ */
+class callback_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a new callback memory resource.
+   *
+   * Constructs a callback memory resource that uses the user-provided callbacks
+   * `allocate_callback` for allocation and `deallocate_callback` for deallocation.
+   *
+   * @param allocate_callback The callback function used for allocation
+   * @param deallocate_callback The callback function used for deallocation
+   * @param allocate_callback_arg Additional context passed to `allocate_callback`.
+   * It is the caller's responsibility to maintain the lifetime of the pointed-to data
+   * for the duration of the lifetime of the `callback_memory_resource`.
+   * @param deallocate_callback_arg Additional context passed to `deallocate_callback`.
+   * It is the caller's responsibility to maintain the lifetime of the pointed-to data
+   * for the duration of the lifetime of the `callback_memory_resource`.
+   */
+  callback_memory_resource(allocate_callback_t allocate_callback,
+                           deallocate_callback_t deallocate_callback,
+                           void* allocate_callback_arg,
+                           void* deallocate_callback_arg) noexcept
```

Comment:
A C++ user of the callback memory resource may not need to jump through these extra hoops to pass in extra context, so it would be better if these parameters could be optional. It would be best if that was also reflected in the signatures `allocate_callback_t` and `deallocate_callback_t`, but at minimum these two constructor arguments should be optional and `allocate_callback_arg` and `deallocate_callback_arg` should be initialized as `nullptr` (of course, this would add more ways in which a user could illegally access memory, but that's already possible with the current definition so I'm not overly concerned that we'd be making things any worse).

Reply:
True, these should just be defaulted to `nullptr`.

Reply:
Done